File: day2/zig/src/main.zig
const std = @import("std");

const INPUT_FILE = "../input.txt";

const Ordering = enum {
    Greater,
    Equal,
    Less,
};

const GameOutcome = enum {
    Win,
    Lose,
    Draw,

    fn from_char(char: u8) GameOutcome {
        return switch (char) {
            'X' => .Lose,
            'Y' => .Draw,
            'Z' => .Win,
            else => unreachable,
        };
    }
};

const Hand = enum {
    Rock,
    Paper,
    Scissors,

    fn from_char(char: u8) Hand {
        return switch (char) {
            'A', 'X' => .Rock,
            'B', 'Y' => .Paper,
            'C', 'Z' => .Scissors,
            else => unreachable,
        };
    }

    fn play(self: Hand, other: Hand) u32 {
        return self.score() + self.outcome_score(other);
    }

    fn outcome_score(self: Hand, other: Hand) u32 {
        return switch (self.cmp(other)) {
            .Greater => 6,
            .Equal => 3,
            .Less => 0,
        };
    }

    fn pivot_hand(self: Hand, outcome: GameOutcome) Hand {
        return switch (self) {
            .Rock => switch (outcome) {
                .Win => .Paper,
                .Draw => .Rock,
                .Lose => .Scissors,
            },
            .Paper => switch (outcome) {
                .Win => .Scissors,
                .Draw => .Paper,
                .Lose => .Rock,
            },
            .Scissors => switch (outcome) {
                .Win => .Rock,
                .Draw => .Scissors,
                .Lose => .Paper,
            },
        };
    }

    fn score(self: Hand) u32 {
        return switch (self) {
            .Rock => 1,
            .Paper => 2,
            .Scissors => 3,
        };
    }

    fn cmp(self: Hand, other: Hand) Ordering {
        return switch (self) {
            .Rock => Hand.cmp_rock(other),
            .Paper => Hand.cmp_paper(other),
            .Scissors => Hand.cmp_scissors(other),
        };
    }

    fn cmp_rock(other: Hand) Ordering {
        return switch (other) {
            .Rock => .Equal,
            .Paper => .Less,
            .Scissors => .Greater,
        };
    }

    fn cmp_paper(other: Hand) Ordering {
        return switch (other) {
            .Rock => .Greater,
            .Paper => .Equal,
            .Scissors => .Less,
        };
    }

    fn cmp_scissors(other: Hand) Ordering {
        return switch (other) {
            .Rock => .Less,
            .Paper => .Greater,
            .Scissors => .Equal,
        };
    }
};

pub fn main() !void {
    var gp = std.heap.GeneralPurposeAllocator(.{ .safety = true }){};
    defer _ = gp.deinit();
    const allocator = gp.allocator();

    var input = try std.fs.cwd().openFile(INPUT_FILE, .{});
    defer input.close();

    var input_meta = try input.metadata();
    var input_size = input_meta.size();

    const input_buffer = try input.readToEndAlloc(allocator, input_size);
    defer allocator.free(input_buffer);

    {
        var score: u32 = 0;
        var iter_line = std.mem.split(u8, input_buffer, "\n");
        while (iter_line.next()) |line| {
            const opponent = Hand.from_char(line[0]);
            const you = Hand.from_char(line[2]);
            score += you.play(opponent);
        }
        std.log.info("total score: {d}", .{score});
    }

    {
        var score: u32 = 0;
        var iter_line = std.mem.split(u8, input_buffer, "\n");
        while (iter_line.next()) |line| {
            const opponent = Hand.from_char(line[0]);
            const outcome = GameOutcome.from_char(line[2]);
            const you = opponent.pivot_hand(outcome);
            score += you.play(opponent);
        }
        std.log.info("total score: {d}", .{score});
    }
}

File: day2/rust/src/main.rs
use std::cmp::Ordering;

const INPUT_FILE: &str = "../input.txt";

fn main() -> Result<(), std::io::Error> {
    part_1()?;
    part_2()?;
    Ok(())
}

enum GameOutcome {
    Win,
    Lose,
    Draw,
}

impl GameOutcome {
    fn from_str(letter: &str) -> Self {
        match letter.chars().next().unwrap() {
            'X' => Self::Lose,
            'Y' => Self::Draw,
            'Z' => Self::Win,
            _ => unreachable!(),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Hand {
    Rock,
    Paper,
    Scissors,
}

impl PartialOrd for Hand {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self == other {
            Some(Ordering::Equal)
        } else {
            Some(match (self, other) {
                (Self::Rock, Self::Scissors)
                | (Self::Paper, Self::Rock)
                | (Self::Scissors, Self::Paper) => Ordering::Greater,
                _ => Ordering::Less,
            })
        }
    }
}

impl Ord for Hand {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap()
    }
}

impl Hand {
    fn from_str(letter: &str) -> Self {
        match letter.chars().next().unwrap() {
            'A' | 'X' => Self::Rock,
            'B' | 'Y' => Self::Paper,
            'C' | 'Z' => Self::Scissors,
            _ => unreachable!(),
        }
    }

    fn score(&self) -> u32 {
        match self {
            Self::Rock => 1,
            Self::Paper => 2,
            Self::Scissors => 3,
        }
    }

    fn pivot_hand(&self, outcome: GameOutcome) -> Self {
        match self {
            Self::Rock => match outcome {
                GameOutcome::Win => Self::Paper,
                GameOutcome::Draw => Self::Rock,
                GameOutcome::Lose => Self::Scissors,
            },
            Self::Paper => match outcome {
                GameOutcome::Win => Self::Scissors,
                GameOutcome::Draw => Self::Paper,
                GameOutcome::Lose => Self::Rock,
            },
            Self::Scissors => match outcome {
                GameOutcome::Win => Self::Rock,
                GameOutcome::Draw => Self::Scissors,
                GameOutcome::Lose => Self::Paper,
            },
        }
    }

    fn play(&self, opponent: Self) -> u32 {
        self.score()
            + match self.cmp(&opponent) {
                Ordering::Greater => 6,
                Ordering::Equal => 3,
                Ordering::Less => 0,
            }
    }
}

fn part_1() -> Result<(), std::io::Error> {
    let input = std::fs::read_to_string(INPUT_FILE)?;
    let total_score = input
        .lines()
        .map(|line| {
            let mut iter = line.split_whitespace();
            let opponent = Hand::from_str(iter.next().unwrap());
            let you = Hand::from_str(iter.next().unwrap());
            you.play(opponent)
        })
        .sum::<u32>();
    println!("total score: {total_score}");
    Ok(())
}

fn part_2() -> Result<(), std::io::Error> {
    let input = std::fs::read_to_string(INPUT_FILE)?;
    let total_score = input
        .lines()
        .map(|line| {
            let mut iter = line.split_whitespace();
            let opponent = Hand::from_str(iter.next().unwrap());
            let outcome = GameOutcome::from_str(iter.next().unwrap());
            let you = opponent.pivot_hand(outcome);
            you.play(opponent)
        })
        .sum::<u32>();
    println!("total score: {total_score}");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hands() {
        let rock = Hand::Rock;
        let paper = Hand::Paper;
        let scissors = Hand::Scissors;

        assert_eq!(rock, Hand::from_str("A"));
        assert_eq!(rock, Hand::from_str("X"));
        assert_eq!(paper, Hand::from_str("B"));
        assert_eq!(paper, Hand::from_str("Y"));
        assert_eq!(scissors, Hand::from_str("C"));
        assert_eq!(scissors, Hand::from_str("Z"));

        assert_eq!(rock.play(rock), 1 + 3);
        assert_eq!(rock.play(paper), 1 + 0);
        assert_eq!(rock.play(scissors), 1 + 6);

        assert_eq!(paper.play(rock), 2 + 6);
        assert_eq!(paper.play(paper), 2 + 3);
        assert_eq!(paper.play(scissors), 2 + 0);

        assert_eq!(scissors.play(rock), 3 + 0);
        assert_eq!(scissors.play(paper), 3 + 6);
        assert_eq!(scissors.play(scissors), 3 + 3);
    }
}
